### DISCUSSION TITLE
(Single instance) Client + SSR + express -> refetching double, bot not in refresh

### DISCUSSION THREAD

**Discussion by remko79** [urql-graphql/urql#1874] [Q&A] [answered]
We have an express server which renders preacts components to string and serves them. Currently we're just trying to get everything server side rendered without any client side hydration (to simplify things). For graphql requests we are using a client with the ssrExchange:

```js
const ssr = ssrExchange({ isClient: false });
const client = createClient({
  url: 'http://localhost:3001/graphql',
  suspense: true, // This activates urql's Suspense mode on the server-side
  exchanges: [
    ssr,
    fetchExchange,
  ],
  fetch,
  preferGetMethod: true,
  requestPolicy: 'network-only',
});
```
'Fetch' is node-fetch.

And the express handler is:
```js
const handleRequest = async (req, res) => {
  const element = (
    <Provider value={client}>
      <Test />
    </Provider>
  );

  // Using `react-ssr-prepass` this prefetches all data
  await prepass(element);
  // This is the usual React SSR rendering code
  const markup = renderToString(element);

  res.send(`
    <html>
      <body>
        <div id="root">${markup}</div>
      </body>
    </html>
  `);
};
```
The Test component is:
```js
import { useQuery } from '@urql/preact';

export default () => {
  const [result] = useQuery({
    query: 'query { article(id: 1) { title } }',
  });

  return (
    <div>test {result.data.article.uid}</div>
  );
};
```
Now 2 things are happening:
1. 2 graphql requests are sent. Is that normal? Can we avoid that somehow?
2. After refreshing the page in the browser no graphql request is made anymore, even when setting the requestPolicy to 'network-only'.

Could someone help me with these issues?

Some remarks: 
* I've set the response headers for the graphql to no-cache and max-age=30
* GET/POST doesn't matter
* I seems to be working if I'm going to create a client for each request like the example in https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/ (so inside the handler) but that seems weird to me and it'll create a big overhead since we'll be getting tons of requests

**Comment by kitten** [comment, marked as answer]
Long story short 😅  I haven't really taken a long look here, but I believe it's the same answer as this one: https://github.com/FormidableLabs/urql/issues/1875

When you say `network-only`, you're using no cache exchange, and you're using suspense, it's like you're telling urql to "not use a cache" three times. This means that duplicate requests may occur since the result is discarded and cannot be retrieved from anywhere. Especially with suspense this is an issue as we're sharing results for as long as suspense is _ongoing_, meaning that once the component is remounted and the effect runs it doesn't have access to the suspended data anymore, since it's not in a cache and it can't be kept around. It can't be kept around since to comply with Suspense and Concurrent React, we cannot keep anything around "inbetween" something _being_ suspended and something remounting after, since that remount may not happen or the effect may never actually run (like a silent unmount)

Edit: Sorry, I forgot to actually write a full response to suggest a solution.

> I seems to be working if I'm going to create a client for each request [...]  but that seems weird to me and it'll create a big overhead since we'll be getting tons of requests

The good news is, it really isn't! ❤️ It's quite cheap and that's pretty much intended. You do want to isolate each SSR request to its own instance / state.

**Reply by remko79** [reply]
Thanks for your quick reply... it works now!

PS. I can't hit the 'Mark as answer' button.. just gives a 422 back on the popup for explanation of 'Marking answers'.... weird stuff
